File: apps/workbench/src/tunnels/adapters/registry.ts
import type { AdapterManifest, BaseAdapter } from '../core/types.ts'

// ── Runtime: adapter manifests + factory ──

const manifests = new Map<string, AdapterManifest>()

export function registerAdapterType(manifest: AdapterManifest): void {
  manifests.set(manifest.type, manifest)
}

export function getAdapterType(type: string): AdapterManifest | undefined {
  return manifests.get(type)
}

export function listAdapterTypes(): Omit<AdapterManifest, 'factory' | 'schema'>[] {
  return Array.from(manifests.values()).map((m) => ({
    type: m.type,
    label: m.label,
    defaultConfig: m.defaultConfig,
  }))
}

export function createAdapter(type: string, config: any): BaseAdapter {
  const manifest = manifests.get(type)
  if (manifest) {
    const parsed = manifest.schema.parse(config)
    return manifest.factory(parsed)
  }
  throw new Error(`Unknown adapter type: ${type}`)
}

export function getDefaultConfig(type: string): Record<string, any> {
  const manifest = manifests.get(type)
  if (manifest) return manifest.defaultConfig
  return {}
}

File: packages/wlya-adapters/Registry.kt
package com.wlya.core.adapters

import com.wlya.core.BaseAdapter

/**
 * Registry of adapter types. Manifests are registered via generated [registerAllAdapters].
 */
data class AdapterManifest(
    val type: String,
    val label: String,
    val defaultConfig: Map<String, Any> = emptyMap(),
    val factory: (config: Map<String, Any>) -> BaseAdapter,
)

object Registry {
    private val manifests = mutableMapOf<String, AdapterManifest>()

    fun register(manifest: AdapterManifest) {
        manifests[manifest.type] = manifest
    }

    fun get(type: String): AdapterManifest? = manifests[type]

    fun list(): List<AdapterManifest> = manifests.values.toList()

    fun createAdapter(type: String, config: Map<String, Any>): BaseAdapter {
        val manifest = manifests[type] ?: throw IllegalArgumentException("Unknown adapter type: $type")
        return manifest.factory(config)
    }

    fun getDefaultConfig(type: String): Map<String, Any> {
        val manifest = manifests[type] ?: throw IllegalArgumentException("Unknown adapter type: $type")
        return manifest.defaultConfig.toMap()
    }
}
